multi/authentication.go

```diff
@@ -45,53 +46,71 @@ var (
 )
 
 func cleanupOldSessions(p *Proxy) {
-	sesssMtx.Lock()
-
-	usernames := make([]string, 0, len(sesss))
-	invalidate := make([]string, 0)
-	// collect the outdated logins
-	for sessionId, data := range sesss {
-		if time.Since(data.Login) > session.SessionTTL {
-			invalidate = append(invalidate, sessionId)
-
-			// audit log of server side logouts
-			if data.User != nil {
-				zap.L().Info(
-					fmt.Sprintf("[AUDIT] Logout '%s' (TTL)",
-						data.User.Username))
-			}
-		} else if data.User != nil {
-			// collect username from the active user sessions
-			usernames = append(usernames, data.User.Username)
-		}
+    sesssMtx.Lock()
+
+    usernames := make([]string, 0, len(sesss))
+    invalidate := make([]string, 0)
+    // Collect the outdated sessions
+    for sessionId, data := range sesss {
+        if time.Since(data.LastActive) > session.SessionTTL {
+            invalidate = append(invalidate, sessionId)
+
+            // Audit log of server side logouts
+            if data.User != nil {
+                zap.L().Info("[AUDIT] Logout", zap.String("user", data.User.Username), zap.String("reason", "TTL"))
+            }
+        } else if data.User != nil {
+            // Collect username from the active user sessions
+            usernames = append(usernames, data.User.Username)
+        }
+    }
+
+    // Invalidate outdated sessions
+    for _, sessionId := range invalidate {
+        delete(sesss, sessionId)
+    }
+
+    sesssMtx.Unlock()
+
+    if p == nil {
+        return
+    }
+
+    // Additional cleanup for user routers
+    for username := range p.r {
+        hasSessionForUser := false
+        for _, loggedinUser := range usernames {
+            if loggedinUser == username {
+                hasSessionForUser = true
+                break
+            }
+        }
+		// delete routers for no longer logged in (LDAP?) users...
+        if !hasSessionForUser && username != router.DefaultRouter {
+            mtx.Lock()
+            delete(p.r, username)
+            mtx.Unlock()
+        }
+    }
+}
 
-	}
-	// invalidate them, lets do also audit log
-	for _, sessionId := range invalidate {
-		delete(sesss, sessionId)
-	}
+func StartSessionCleanupScheduler(p *Proxy) {
+    ticker := time.NewTicker(30 * time.Minute)
```

Comment:
Do we need this? maybe i'm wrong, but it looks to me that we are calling cleanup on every request inside RPCAuthMiddleware

Reply:
ah, right, cleanupOldSessions is removed from RPCAuthMiddleware